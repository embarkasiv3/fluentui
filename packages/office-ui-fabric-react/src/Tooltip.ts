export * from './packages/react-fundamentals/Tooltip';
