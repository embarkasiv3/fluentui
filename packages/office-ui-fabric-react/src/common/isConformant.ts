import { isConformant as isConformantBase, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: IsConformantOptions) {
  const defaultOptions = {
    disabledTests: ['has-docblock'],
  };

  isConformantBase(defaultOptions, testInfo);
}
