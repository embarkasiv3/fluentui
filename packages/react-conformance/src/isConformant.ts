import { TestObject, TestingOptions } from './types';
import { withCustomConfig } from 'react-docgen-typescript';
import * as path from 'path';
import { defaultTests } from './defaultTests';

export function isConformant(testInfo: TestingOptions, extraTests?: TestObject) {
  const { componentPath, displayName } = testInfo;
  const tsconfigPath = path.join(process.cwd(), 'tsconfig.json');
  const parser = withCustomConfig(tsconfigPath, {});
  const components = parser.parse(componentPath);
  const mainComponents = components.filter(comp => comp.displayName === displayName);

  if (mainComponents.length === 1) {
    const componentInfo = mainComponents[0];

    for (const test of Object.keys(defaultTests)) {
      defaultTests[test](componentInfo, testInfo);
    }
    if (extraTests) {
      for (const test of Object.keys(extraTests)) {
        extraTests[test](componentInfo, testInfo);
      }
    }
  } else {
    if (components.length === 0) {
      throw new Error('No exported components in path: ' + componentPath);
    } else {
      throw new Error(
        `No component with name '${displayName}' was found at ${componentPath}. ` +
          `These are the exported component names: ${components.map(component => component.displayName).join(', ')}`,
      );
    }
  }
}
