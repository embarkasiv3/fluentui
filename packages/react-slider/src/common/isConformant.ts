import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: IsConformantOptions) {
  const defaultOptions = {
    disabledTests: ['has-docblock', 'kebab-aria-attributes'],
  };

  baseIsConformant(defaultOptions, testInfo);
}
