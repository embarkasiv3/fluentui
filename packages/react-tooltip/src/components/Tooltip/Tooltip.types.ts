import * as React from 'react';
import { Position, Alignment } from '@fluentui/react-positioning';
import { ComponentProps, ComponentState, ShorthandProps } from '@fluentui/react-utilities';

/**
 * {@docCategory Tooltip}
 */
export interface TooltipProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /**
   * How to position the tooltip relative to the target element. This is a "best effort" placement,
   * but the tooltip may be flipped to the other side if there is not enough room.
   *
   * @defaultvalue above
   */
  position?: Position;

  /**
   * How to align the tooltip along the edge of the target element.
   *
   * @defaultvalue center
   */
  align?: Alignment;

  /**
   * Color variant with a subtle look
   */
  subtle?: boolean;

  /**
   * Do not render an arrow pointing to the target element
   */
  noArrow?: boolean;

  /**
   * Distance between the tooltip and the target element, in pixels
   *
   * @defaultvalue 4
   */
  offset?: number;

  /**
   * The arrow that points to the target element. This will be rendered by default unless `noArrow` is specified.
   */
  arrow?: ShorthandProps<React.HTMLAttributes<HTMLElement> & React.RefAttributes<HTMLElement>>;

  /**
   * Imperative handle to show and hide the tooltip
   */
  componentRef?: React.Ref<TooltipImperativeHandle>;
}

/**
 * {@docCategory Tooltip}
 */
export interface TooltipImperativeHandle {
  /**
   * Show the tooltip, pointing to the target element
   */
  show: (target: HTMLElement) => void;

  /**
   * Hide the tooltip
   */
  hide: () => void;
}

/**
 * Names of the shorthand properties in TooltipProps
 * {@docCategory Tooltip}
 */
export type TooltipShorthandProps = 'arrow';

/**
 * Names of TooltipProps that have a default value in useTooltip
 * {@docCategory Tooltip}
 */
export type TooltipDefaultedProps = 'position' | 'align' | 'offset';

/**
 * {@docCategory Tooltip}
 */
export type TooltipState = ComponentState<
  React.Ref<HTMLElement>,
  TooltipProps & {
    visible: boolean;
  },
  TooltipShorthandProps,
  TooltipDefaultedProps
>;
