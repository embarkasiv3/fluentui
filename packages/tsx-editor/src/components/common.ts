/** Font family for code snippets/editors */
export const codeFontFamily = 'Monaco, Menlo, Consolas, "Droid Sans Mono", "Inconsolata", "Courier New", monospace';
