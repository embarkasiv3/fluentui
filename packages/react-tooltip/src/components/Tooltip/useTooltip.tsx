import * as React from 'react';
import { makeMergeProps, resolveShorthandProps, useMergedRefs } from '@fluentui/react-utilities';
import { usePopper } from '@fluentui/react-positioning';
import { useTheme } from '@fluentui/react-theme-provider';
import { TooltipProps, TooltipShorthandProps, TooltipState } from './Tooltip.types';
import { arrowHeight, tooltipBorderRadius } from './useTooltipStyles';

/**
 * Names of the shorthand properties in TooltipProps
 * {@docCategory Tooltip}
 */
export const tooltipShorthandProps: TooltipShorthandProps[] = ['arrow'];

const mergeProps = makeMergeProps<TooltipState>({ deepMerge: tooltipShorthandProps });

/**
 * Create the state required to render Tooltip.
 *
 * The returned state can be modified with hooks such as useTooltipStyles,
 * before being passed to renderTooltip.
 *
 * @param props - props from this instance of Tooltip
 * @param ref - reference to root HTMLElement of Tooltip
 * @param defaultProps - (optional) default prop values provided by the implementing type
 *
 * {@docCategory Tooltip}
 */
export const useTooltip = (
  props: TooltipProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: TooltipProps,
): TooltipState => {
  const [target, setTarget] = React.useState<HTMLElement | null>(null);

  const state = mergeProps(
    {
      ref,
      as: 'div',
      role: 'tooltip',
      position: 'above',
      align: 'center',
      offset: 4,
      arrow: { as: 'div', children: '' },
      visible: !!target,
    },
    defaultProps && resolveShorthandProps(defaultProps, tooltipShorthandProps),
    resolveShorthandProps(props, tooltipShorthandProps),
  );

  if (state.noArrow) {
    state.arrow = undefined;
  }

  React.useImperativeHandle(
    state.componentRef,
    () => ({
      show: t => setTarget(t),
      hide: () => setTarget(null),
    }),
    [],
  );

  const theme = useTheme();

  const { targetRef, containerRef, arrowRef } = usePopper({
    enabled: state.visible,
    position: state.position,
    align: state.align,
    offset: [0, state.offset + (state.noArrow ? 0 : arrowHeight)],
    arrowPadding: theme?.global ? 2 * parseInt(tooltipBorderRadius(theme), 10) : 0,
  });

  targetRef.current = target;
  state.ref = useMergedRefs(state.ref, containerRef);

  const mergedArrowRef = useMergedRefs(state.arrow?.ref, arrowRef);
  if (state.arrow) {
    state.arrow.ref = mergedArrowRef;
  }

  return state;
};
