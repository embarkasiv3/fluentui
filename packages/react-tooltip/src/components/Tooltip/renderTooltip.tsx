import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { TooltipState } from './Tooltip.types';
import { tooltipShorthandProps } from './useTooltip';

/**
 * Render the final JSX of Tooltip
 * {@docCategory Tooltip}
 */
export const renderTooltip = (state: TooltipState) => {
  const { slots, slotProps } = getSlots(state, tooltipShorthandProps);

  return (
    <slots.root {...slotProps.root}>
      <slots.arrow {...slotProps.arrow} />
      {state.children}
    </slots.root>
  );
};
