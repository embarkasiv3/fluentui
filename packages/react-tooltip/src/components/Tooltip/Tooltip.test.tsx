import * as React from 'react';
import { Tooltip } from './Tooltip';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';

describe('Tooltip', () => {
  isConformant({
    Component: Tooltip,
    displayName: 'Tooltip',
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for Tooltip in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<Tooltip>Default Tooltip</Tooltip>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
