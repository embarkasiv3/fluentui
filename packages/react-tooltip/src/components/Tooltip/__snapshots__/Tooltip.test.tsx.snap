// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders a default state 1`] = `
<div
  className=""
  role="tooltip"
>
  <div
    className=""
  >
    
  </div>
  Default Tooltip
</div>
`;
